cluster-autoscaler/cloudprovider/clusterapi/clusterapi_nodegroup.go

```diff
@@ -308,14 +308,39 @@ func (ng *nodegroup) TemplateNodeInfo() (*schedulerframework.NodeInfo, error) {
 	node.Status.Allocatable = capacity
 	node.Status.Conditions = cloudprovider.BuildReadyConditions()
 	node.Spec.Taints = ng.scalableResource.Taints()
-	node.Labels = cloudprovider.JoinStringMaps(ng.scalableResource.Labels(), buildGenericLabels(nodeName))
+
+	node.Labels, err = ng.buildTemplateLabels(nodeName)
+	if err != nil {
+		return nil, err
+	}
 
 	nodeInfo := schedulerframework.NewNodeInfo(cloudprovider.BuildKubeProxy(ng.Name()))
 	nodeInfo.SetNode(&node)
 
 	return nodeInfo, nil
 }
 
+func (ng *nodegroup) buildTemplateLabels(nodeName string) (map[string]string, error) {
+	labels := cloudprovider.JoinStringMaps(ng.scalableResource.Labels(), buildGenericLabels(nodeName))
+
+	nodes, err := ng.Nodes()
+	if err != nil {
+		return nil, err
+	}
+
+	if len(nodes) > 0 {
+		node, err := ng.machineController.findNodeByProviderID(normalizedProviderString(nodes[0].Id))
+		if err != nil {
+			return nil, err
+		}
+
+		if node != nil {
+			labels = cloudprovider.JoinStringMaps(labels, extractNodeLabels(node))
+		}
+	}
```

Comment:
i'm slightly concerned about this clause because it seems like it would be possible for a user to add a label to a node which is not on the machineset, and then we would return that as part of the labels for the node group. this could lead to situations where the autoscaler simulator is evaluating labels that do not exist on the machineset. i'm wondering if this might have implications for how things like `--balance-similar-node-groups` could work in edge cases.

i do see that the `extractNodeLables` does not copy all labels, so this might be more of a theoretical question.

Reply:
The `extractNodeLabels` deliberately only copies a number of well known labels deliberately so that users can't just roll their own random labels on a node, I don't think we'd be able to reliably copy random labels and guarantee the scaled up node would contain those labels. I think with the current implementation we should be safe.

Reply:
makes sense, thanks!